File: app/src/main/java/bassamalim/hidaya/activities/TelawatCollectionActivity.java
package bassamalim.hidaya.activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Location;
import android.os.Bundle;
import android.view.Window;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.preference.PreferenceManager;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.Objects;

import bassamalim.hidaya.R;
import bassamalim.hidaya.databinding.ActivityCollectionTelawatBinding;
import bassamalim.hidaya.enums.ListType;
import bassamalim.hidaya.fragments.TelawatFragment;
import bassamalim.hidaya.helpers.Keeper;
import bassamalim.hidaya.other.Utils;

public class TelawatCollectionActivity extends FragmentActivity {

    private ActivityCollectionTelawatBinding binding;
    private FragmentStateAdapter adapter;
    private ViewPager2 viewPager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.onActivityCreateSetTheme(this);
        binding = ActivityCollectionTelawatBinding.inflate(getLayoutInflater());
        getWindow().requestFeature(Window.FEATURE_ACTION_BAR);
        setContentView(binding.getRoot());

        viewPager = findViewById(R.id.telawat_pager);
        adapter = new TAdapter(this);
        viewPager.setAdapter(adapter);

        TabLayout tabLayout = findViewById(R.id.tab_layout);

        String[] tabs = new String[] {"الكل", "المفضلة", "المحملة"};
        new TabLayoutMediator(tabLayout, viewPager,
                (tab, position) -> tab.setText(tabs[position])
        ).attach();
    }

    @Override
    protected void onResume() {
        super.onResume();
        setupContinue();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        if (viewPager.getCurrentItem() == 0) {
            // If the user is currently looking at the first step, allow the system to handle the
            // Back button. This calls finish() on this activity and pops the back stack.
            super.onBackPressed();

            if (isTaskRoot()) {
                Intent intent = new Intent(this, MainActivity.class);
                Location location = new Keeper(this).retrieveLocation();
                intent.putExtra("located", location != null);
                intent.putExtra("location", location);
                startActivity(intent);
                finish();
            }
        }
        else {
            // Otherwise, select the previous step.
            viewPager.setCurrentItem(viewPager.getCurrentItem() - 1);
        }
    }

    private void setupContinue() {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(this);
        String text = pref.getString("last_played_text", "");

        if (text.length() == 0)
            text = "لم يتم تشغيل أي تلاوة بعد";
        else {
            text = "آخر تشغيل: " + text;

            binding.continueListening.setOnClickListener(v -> {
                String lastMediaId = pref.getString("last_played_media_id", "");

                Intent intent = new Intent(this, TelawatClient.class);
                intent.setAction("continue");
                intent.putExtra("media_id", lastMediaId);
                startActivity(intent);
            });
        }

        binding.continueListening.setText(text);
    }
}

class TAdapter extends FragmentStateAdapter {

    public TAdapter(FragmentActivity fragment) {
        super(fragment);
    }

    @NonNull @Override
    public Fragment createFragment(int position) {
        ListType type;

        if (position == 0)
            type = ListType.All;
        else if (position == 1)
            type = ListType.Favorite;
        else
            type = ListType.Downloaded;

        return new TelawatFragment(type);
    }

    @Override
    public int getItemCount() {
        return 3;
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/TelawatCollectionActivity.kt
package bassamalim.hidaya.activities

import android.content.Intent
import android.location.Location
import android.os.Build
import android.os.Bundle
import androidx.annotation.RequiresApi
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.preference.PreferenceManager
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import bassamalim.hidaya.R
import bassamalim.hidaya.databinding.ActivityCollectionTelawatBinding
import bassamalim.hidaya.enums.ListType
import bassamalim.hidaya.screens.TelawatFragment
import bassamalim.hidaya.helpers.Keeper
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.DBUtils
import bassamalim.hidaya.utils.PrefUtils
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator

@RequiresApi(api = Build.VERSION_CODES.O)
class TelawatCollectionActivity : FragmentActivity() {

    private lateinit var binding: ActivityCollectionTelawatBinding
    private lateinit var adapter: FragmentStateAdapter
    private lateinit var viewPager: ViewPager2

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ActivityUtils.myOnActivityCreated(this)
        binding = ActivityCollectionTelawatBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.home.setOnClickListener { finish() }

        viewPager = findViewById(R.id.telawat_pager)
        adapter = TAdapter(this)
        viewPager.adapter = adapter

        val tabLayout = findViewById<TabLayout>(R.id.tab_layout)
        val tabs = arrayOf(getString(R.string.all), getString(R.string.favorite), getString(R.string.downloaded))
        TabLayoutMediator(tabLayout, viewPager) { tab: TabLayout.Tab, position: Int ->
            tab.text = tabs[position]
        }.attach()
    }

    override fun onResume() {
        super.onResume()
        setupContinue()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        if (viewPager.currentItem == 0) {
            // If the user is currently looking at the first step, allow the system to handle the
            // Back button. This calls finish() on this activity and pops the back stack.
            super.onBackPressed()

            if (isTaskRoot) {
                val intent = Intent(this, MainActivity::class.java)
                val location: Location? = Keeper(this).retrieveLocation()
                intent.putExtra("located", location != null)
                intent.putExtra("location", location)
                startActivity(intent)
                finish()
            }
        }
        else {
            // Otherwise, select the previous step.
            viewPager.currentItem = viewPager.currentItem - 1
        }
    }

    private fun setupContinue() {
        val pref = PreferenceManager.getDefaultSharedPreferences(this)
        val mediaId = pref.getString("last_played_media_id", "")!!

        if (mediaId.isEmpty()) binding.continueListening.text = getString(R.string.no_last_play)
        else {
            val db = DBUtils.getDB(this)

            val reciterId = mediaId.substring(0, 3).toInt()
            val versionId = mediaId.substring(3, 5).toInt()
            val suraIndex = mediaId.substring(5).toInt()

            val suraName =
                if (PrefUtils.getLanguage(this, pref) == "en")
                    db.suarDao().getNameEn(suraIndex)
                else db.suarDao().getName(suraIndex)
            val reciterName = db.telawatRecitersDao().getName(reciterId)
            val rewaya = db.telawatVersionsDao().getVersion(reciterId, versionId).getRewaya()

            val text = "${getString(R.string.last_play)}: " +
                    "${getString(R.string.sura)} $suraName ${getString(R.string.for_reciter)} $reciterName ${getString(R.string.in_rewaya_of)} $rewaya"
            binding.continueListening.text = text

            binding.continueListening.setOnClickListener {
                val intent = Intent(this, TelawatClient::class.java)
                intent.action = "continue"
                intent.putExtra("media_id", mediaId)
                startActivity(intent)
            }
        }
    }

}

@RequiresApi(api = Build.VERSION_CODES.O)
internal class TAdapter(fragment: FragmentActivity?) : FragmentStateAdapter(fragment!!) {

    override fun createFragment(position: Int): Fragment {
        val type = when (position) {
            0 -> ListType.All
            1 -> ListType.Favorite
            else -> ListType.Downloaded
        }
        return TelawatFragment(type)
    }

    override fun getItemCount(): Int {
        return 3
    }

}